core/src/main/python/wlsdeploy/util/variables.py

```diff
@@ -438,19 +438,33 @@ def _report_token_issue(message_key, method_name, model_context, *args):
         raise ex
 
 
-def substitute_key(text, variables):
+def substitute_key(text, variables, model_context):
     """
     Substitute any @@PROP values in the text and return.
     If the corresponding variable is not found, leave the @@PROP value in place.
     :param text: the text to be evaluated
     :param variables: the variable map
     :return: the substituted text value
     """
-    matches = _property_pattern.findall(text)
+    method_name = "substitute_key"
+
+    if variables is not None:
+        matches = _property_pattern.findall(text)
+        for token, key in matches:
+            if key in variables:
+                value = variables[key]
+                text = text.replace(token, value)
+
+    matches = _environment_pattern.findall(text)
     for token, key in matches:
-        if key in variables:
-            value = variables[key]
-            text = text.replace(token, value)
+        # log, or throw an exception if key is not found.
+        if not os.environ.has_key(key):
+            _report_token_issue('WLSDPLY-01737', method_name, model_context, key)
+            continue
```

Comment:
Why not just leave this in the merged model as `@@ENV:...@@` if it can't be resolved, similar to the way `@@PROP` is done here?
I think it's reasonable (in general) to merge models with unresolved ENV or PROP, and log/throw any problems later when the model is used.
The changes to method `_get_merge_match_key` should take care of merging names when the ENV value is matched to a fixed value in the other model (WDT-552).

Reply:
We can do that.